File: backend/app/schemas/analytics.py
from typing import List, Optional
from pydantic import BaseModel, Field


class FunnelStageData(BaseModel):
    stageId: str
    label: str
    count: int
    percentageOfTop: float
    conversionFromPrevious: float
    dropOffCount: int
    dropOffPercentage: float


class FunnelInsightText(BaseModel):
    highlight: str
    dropOffDetail: str


class ExecutiveMetric(BaseModel):
    id: str
    label: str
    value: str
    trendPercentage: float = 0.0
    trendDirection: str = "neutral"
    comparisonLabel: str = "vs previous period"
    context: Optional[str] = None


class IntentDistributionCount(BaseModel):
    count: int
    percentage: float


class IntentLevelDistribution(BaseModel):
    veryHigh: IntentDistributionCount
    high: IntentDistributionCount
    medium: IntentDistributionCount
    low: IntentDistributionCount


class CallPerformanceData(BaseModel):
    totalCalls: int
    avgDurationSeconds: int
    qualifiedCalls: int
    interestedCalls: int
    followUpsCreated: int
    qualificationRate: float
    interestedRate: float


class CallOutcomeItem(BaseModel):
    outcome: str
    label: str
    count: int
    percentage: float
    colorClass: str = "bg-primary text-primary"
    badgeStyle: str = "bg-primary-muted text-primary border-primary/30"


class StageConversionMetric(BaseModel):
    id: str
    title: str
    fromStage: str
    toStage: str
    rate: float
    trendPercentage: float = 0.0
    benchmark: float = 30.0
    status: str = "on_track"


class SourcePerformanceItem(BaseModel):
    sourceKey: str
    sourceLabel: str
    platformIconName: Optional[str] = None
    discoveredLeads: int
    highIntentLeads: int
    qualifiedLeads: int
    qualificationRate: float
    avgIntentScore: float


class IndustryPerformanceItem(BaseModel):
    industry: str
    discoveredLeads: int
    highIntentCount: int
    qualifiedCount: int
    qualificationRate: float
    totalEstimatedValue: str


class AnalyticsInsight(BaseModel):
    id: str
    category: str
    categoryLabel: str
    title: str
    evidence: str
    whyItMatters: str
    recommendedAction: str
    ctaLabel: Optional[str] = None
    ctaTarget: Optional[str] = None


class TrendChartPoint(BaseModel):
    label: str
    value: int
    highlight: bool = False


class AnalyticsMetricsResponse(BaseModel):
    dateRange: str = "30d"
    discoveredCount: int
    contactedCount: int
    interestedCount: int
    convertedCount: int
    conversionRate: float
    funnelStages: List[FunnelStageData]
    funnelInsightText: FunnelInsightText
    executiveMetrics: List[ExecutiveMetric]
    intentDistribution: IntentLevelDistribution
    callPerformance: CallPerformanceData
    callOutcomes: List[CallOutcomeItem]
    conversions: List[StageConversionMetric] = Field(default_factory=list)
    sourcePerformance: List[SourcePerformanceItem] = Field(default_factory=list)
    industryPerformance: List[IndustryPerformanceItem] = Field(default_factory=list)
    insights: List[AnalyticsInsight] = Field(default_factory=list)
    trendChartData: List[TrendChartPoint] = Field(default_factory=list)

File: src/types/analytics.ts
export type DateRangePreset = 'today' | '7d' | '15d' | '30d' | '90d' | 'custom';

export interface ExecutiveMetric {
  id: string;
  label: string;
  value: string | number;
  trendPercentage: number;
  trendDirection: 'up' | 'down' | 'neutral';
  comparisonLabel: string; // e.g. "vs previous period"
  context?: string;
}

export interface FunnelStageData {
  stageId: 'discovered' | 'high_intent' | 'contacted' | 'qualified' | 'meeting' | 'opportunity';
  label: string;
  count: number;
  percentageOfTop: number; // Conversion relative to Discovered
  conversionFromPrevious: number; // Conversion relative to previous stage
  dropOffCount: number;
  dropOffPercentage: number;
}

export interface IntentLevelDistribution {
  veryHigh: { count: number; percentage: number };
  high: { count: number; percentage: number };
  medium: { count: number; percentage: number };
  low: { count: number; percentage: number };
}

export interface StageConversionMetric {
  id: string;
  title: string;
  fromStage: string;
  toStage: string;
  rate: number; // Percentage, e.g. 50.0
  trendPercentage: number;
  benchmark: number;
  status: 'above' | 'on_track' | 'below';
}

export interface CallPerformanceData {
  totalCalls: number;
  avgDurationSeconds: number; // e.g. 222 (3m 42s)
  qualifiedCalls: number;
  interestedCalls: number;
  followUpsCreated: number;
  qualificationRate: number; // e.g. 50%
  interestedRate: number; // e.g. 22.2%
}

export interface CallOutcomeItem {
  outcome: 'QUALIFIED' | 'INTERESTED' | 'FOLLOW_UP' | 'NURTURE' | 'NOT_INTERESTED' | 'NO_ANSWER' | 'FAILED';
  label: string;
  count: number;
  percentage: number;
  colorClass: string;
  badgeStyle: string;
}

export interface SourcePerformanceItem {
  sourceKey: string;
  sourceLabel: string;
  platformIconName?: string;
  discoveredLeads: number;
  highIntentLeads: number;
  qualifiedLeads: number;
  qualificationRate: number; // Percentage e.g. 38%
  avgIntentScore: number;
}

export interface IndustryPerformanceItem {
  industry: string;
  discoveredLeads: number;
  highIntentCount: number;
  qualifiedCount: number;
  qualificationRate: number; // Percentage
  totalEstimatedValue: string; // e.g. "₹42.5L"
}

export interface AnalyticsInsight {
  id: string;
  category: 'HIGH_PERFORMING_SOURCE' | 'CALL_CONVERSION' | 'FUNNEL_DROPOFF' | 'INTENT_VELOCITY';
  categoryLabel: string;
  title: string;
  evidence: string;
  whyItMatters: string;
  recommendedAction: string;
  ctaLabel?: string;
  ctaTarget?: string; // route e.g. "/leads/discover" or "/leads"
}

export interface SalesAnalyticsDataset {
  dateRange: DateRangePreset;
  executiveMetrics: ExecutiveMetric[];
  funnelStages: FunnelStageData[];
  funnelInsightText: {
    highlight: string;
    dropOffDetail: string;
  };
  intentDistribution: IntentLevelDistribution;
  conversions: StageConversionMetric[];
  callPerformance: CallPerformanceData;
  callOutcomes: CallOutcomeItem[];
  sourcePerformance: SourcePerformanceItem[];
  industryPerformance: IndustryPerformanceItem[];
  insights: AnalyticsInsight[];
  trendChartData?: { label: string; value: number; highlight?: boolean }[];
}
